README
f.lux indicator applet
Better lighting for your computer

f.lux indicator applet is an indicator applet to control xflux, an application
that makes the color of your computer's display adapt to the time of day, warm
at nights and like sunlight during the day



debian/changelog
# -*- text -*-
fluxgui (2.0.1~pre) wily; urgency=low

 -- Nathan Collins <nathan.collins@gmail.com>  <add date here when releasing>

fluxgui (2.0) wily; urgency=low

  * add support for Redshift as the backend
    https://github.com/xflux-gui/fluxgui/pull/143.

 -- Nathan Collins <nathan.collins@gmail.com>  Wed, 28 Sep 2022 20:51:26 +0800

 
fluxgui (1.2.0) wily; urgency=low

  * port fluxgui to GTK+3 and Python 3
    https://github.com/xflux-gui/fluxgui/pull/112.
  * make fluxgui Gnome autostart work for locally installed fluxgui,
    https://github.com/xflux-gui/fluxgui/pull/89.
  * update icons: https://github.com/xflux-gui/fluxgui/issues/85,
    https://github.com/xflux-gui/fluxgui/pull/86.
  * fix https:/github.com/xflux-gui/fluxgui/issues/79: tray icon
    name was incorrect for some desktop environments.

 -- Nathan Collins <nathan.collins@gmail.com>  Fri, 03 May 2019 07:41:03 -0700

fluxgui (1.1.10) wily; urgency=low

  * fix https://github.com/xflux-gui/xflux-gui/issues/51: some of
    the the color temps I added in 1.1.9 were not actually supported :P
  * fix gtk-icon-theme related warnings. These were innocuous, but
    mentioned in most bug reports.
  * fix https://github.com/xflux-gui/xflux-gui/issues/35: screen color
    was not reset to normal when closing fluxgui.
  * fix https://github.com/xflux-gui/xflux-gui/issues/38: remove confusing
    GUI element.
  * fix https://github.com/xflux-gui/xflux-gui/issues/16: previewing the
    already chosen color during the daytime didn't work.

 -- Nathan Collins <nathan.collins@gmail.com>  Fri, 09 Jun 2017 17:17:32 -0700

fluxgui (1.1.9) wily; urgency=low

  * remove pid file / lock file. We have several bug reports about fluxgui
    failing to start due to stale pid files, but I doubt anyone has the problem
    of accidentally starting multiple copies of fluxgui. We can revisit this
    if it turns out not having the pid file causes problems.
  * fix bug where preview did not work for already chosen temperature.
  * add more color options to the preferences GUI. Some of these options
    used to be supported in older versions, but had gone away at some point.

 -- Nathan Collins <nathan.collins@gmail.com>  Sat, 23 Jul 2016 16:28:45 -0700

fluxgui (1.1.8) lucid; urgency=low

  * add in python-pexpect as a dependency 

 -- Kilian Valkhof <kilian@kilianvalkhof.com>  Wed, 08 Sep 2010 12:12:16 +0200

fluxgui (1.1.7-0ubuntu1) lucid; urgency=low

  * handle xflux closing more gracefully (thanks to axonxorz) 

 -- Kilian Valkhof <kilian@kilianvalkhof.com>  Sat, 31 Jul 2010 15:29:01 +0200

fluxgui (1.1.6-0ubuntu1) lucid; urgency=low

  * make the preview button preview the selected value instead of current
  * update xflux
  * improved KDE support (needs testing)

 -- Kilian Valkhof <kilian@ktop3>  Thu, 29 Jul 2010 14:32:55 +0200

fluxgui (1.1.5-0ubuntu1) lucid; urgency=low

  * remove any stray lock files if xflux can't be started (by seancron) 

 -- Kilian Valkhof <kilian@kilianvalkhof.com>  Mon, 26 Jul 2010 23:15:51 +0200

fluxgui (1.1.4-0ubuntu1) lucid; urgency=low

  * fix problems with the first start of the application when there are no settings yet 

 -- Kilian Valkhof <kilian@kilianvalkhof.com>  Mon, 26 Jul 2010 17:46:41 +0200

fluxgui (1.1.3-0ubuntu1) lucid; urgency=low

  * clean up properly after a keyboardinterrupt 

 -- Kilian Valkhof <kilian@kilianvalkhof.com>  Mon, 26 Jul 2010 14:05:51 +0200

fluxgui (1.1.2-0ubuntu1) lucid; urgency=low

  * add a close button to the preferences pane 

 -- Kilian Valkhof <kilian@kilianvalkhof.com>  Mon, 26 Jul 2010 12:08:10 +0200

fluxgui (1.1.1-0ubuntu1) lucid; urgency=low

  * update some debian settings

 -- Kilian Valkhof <kilian@kilianvalkhof.com>  Sun, 25 Jul 2010 14:18:19 +0200

fluxgui (1.1.0-0ubuntu1) lucid; urgency=low

  * prevent fluxgui from starting more than once
  * implement a preview button
  * implement an autostart function

 -- Kilian Valkhof <kilian@kilianvalkhof.com>  Thu, 15 Jul 2010 21:08:59 +0200

fluxgui (1.0.0-0ubuntu1) lucid; urgency=low

  * gnome indicator applet for xflux
  * own logo
  * starts xflux automatically
  * adapts icon to lucid style
  * allows on/off toggling of xflux
  * saves the latitude/zipcode settings in gconf
  * opens preferences window if latitude/zipcode is not yet set

 -- Kilian Valkhof <kilian@kilianvalkhof.com>  Tue, 15 Jun 2010 22:14:58 +0200



src/fluxgui/settings.py
import os
from gi.repository import Gio
from xdg.DesktopEntry import DesktopEntry
from fluxgui.exceptions import DirectoryCreationError

################################################################
# Color temperatures.

# The color options available in the color preferences dropdown in the
# "Preferences" GUI are defined in ./preferences.glade. Choosing a
# preference in the GUI returns a number, with 0 for the first choice,
# 1 for the second choice, etc.
default_temperature = '3400'
off_temperature = '6500'
temperatures = [
    '2000',  # The minimum supported by flux; see https://github.com/xflux-gui/xflux-gui/issues/51
    '2300',
    '2700',
    '3400',  # The 'default_temperature' needs to be one of the options!
    '4200',
    '5000',
    # The "off temperature" is not one of the menu choices, but
    # the previous code included it, so @ntc2 is leaving it in
    # without understanding why ...
    #
    # TODO(ntc2): understand why this entry is in the list, and remove
    # it if possible.
    off_temperature]


def key_to_temperature(key):
    """The inverse of 'temperature_to_key'.

    """
    # The old version of this code supported a special key "off". We
    # now map all unknown keys to "off", but I don't understand what
    # the "off" value is for.
    #
    # TODO(ntc2): figure out what the "off" value is for.
    try:
        return temperatures[key]
    except IndexError:
        return off_temperature


def temperature_to_key(temperature):
    """Convert a temperature like "3400" to a Glade/GTK menu value like
    "1" or "off".

    """
    for i, t in enumerate(temperatures):
        if t == temperature:
            return i
    # For invalid temperatures -- which should be impossible ? --
    # return the number corresponding to the off temperature. Perhaps
    # we could also return "off" here? But I have no idea how this
    # code is even triggered.
    return len(temperatures) - 1

################################################################


class Settings(object):

    def __init__(self):
        # You can use
        #
        #     gsettings [--schema-dir .] list-recursively apps.fluxgui
        #
        # to see current settings on command line. The '--schema-dir'
        # is necessary if you're using a non-standard schema dir,
        # e.g. when running fluxgui from the repo without installing
        # it.
        self.settings = Gio.Settings.new('apps.fluxgui')

        self._color = self.settings.get_string("colortemp")
        self._autostart = self.settings.get_boolean("autostart")
        self._latitude = self.settings.get_string("latitude")
        self._longitude = self.settings.get_string("longitude")
        self._zipcode = self.settings.get_string("zipcode")
        self._use_redshift = self.settings.get_boolean("useredshift")
        self._use_xflux = self.settings.get_boolean("usexflux")

        self.has_set_prefs = True
        if not self._latitude and not self._zipcode:
            self.has_set_prefs = False
            self._zipcode = '90210'

        # After an upgrade to fluxgui where the color options change,
        # the color setting may no longer be one of the menu
        # options. In this case, we reset to the default night time
        # temp.
        if self._color not in temperatures:
            self.color = default_temperature
        else:
            self.color = self._color

    def xflux_settings_dict(self):
        d = {
            'color': self.color,
            'latitude': self.latitude,
            'longitude': self.longitude,
            'zipcode': self.zipcode,
            'pause_color': off_temperature
        }
        return d

    def redshift_settings_dict(self):
        return {
            'color': self.color,
            'latitude': self.latitude,
            'longitude': self.longitude,
            'pause_color': off_temperature
        }

    def _get_color(self):
        return self._color

    def _set_color(self, value):
        self._color = value
        self.settings.set_string("colortemp", value)

    def _get_latitude(self):
        return self._latitude

    def _set_latitude(self, value):
        self._latitude = value
        self.settings.set_string("latitude", value)

    def _get_longitude(self):
        return self._longitude

    def _set_longitude(self, value):
        self._longitude = value
        self.settings.set_string("longitude", value)

    def _get_zipcode(self):
        return self._zipcode

    def _set_zipcode(self, value):
        self._zipcode = value
        self.settings.set_string("zipcode", value)

    def _get_autostart(self):
        return self._autostart

    def _set_autostart(self, value):
        self._autostart = value
        self.settings.set_boolean("autostart", self._autostart)
        if self._autostart:
            self._create_autostarter()
        else:
            self._delete_autostarter()

    def _set_xflux(self, value):
        self._use_xflux = value
        self.settings.set_boolean("usexflux", value)

    def _get_xflux(self):
        return self._use_xflux

    def _set_redshift(self, value):
        self._use_redshift = value
        self.settings.set_boolean("useredshift", value)

    def _get_redshift(self):
        return self._use_redshift

    color = property(_get_color, _set_color)
    latitude = property(_get_latitude, _set_latitude)
    longitude = property(_get_longitude, _set_longitude)
    zipcode = property(_get_zipcode, _set_zipcode)
    autostart = property(_get_autostart, _set_autostart)
    use_redshift = property(_get_redshift, _set_redshift)
    use_xflux = property(_get_xflux, _set_xflux)

    # autostart code copied from AWN

    def _get_autostart_file_path(self):
        autostart_dir = os.path.join(os.environ['HOME'], '.config',
                                     'autostart')
        return os.path.join(autostart_dir, 'fluxgui.desktop')

    def _create_autostarter(self):
        autostart_file = self._get_autostart_file_path()
        autostart_dir = os.path.dirname(autostart_file)

        if not os.path.isdir(autostart_dir):
            # create autostart dir
            try:
                os.mkdir(autostart_dir)
            except DirectoryCreationError as e:
                print("Creation of autostart dir failed, please make it yourself: {}".format(autostart_dir))
                raise e

        if not os.path.isfile(autostart_file):
            # create autostart entry
            starter_item = DesktopEntry(autostart_file)
            starter_item.set('Name', 'f.lux indicator applet')
            # Use the user's shell to start 'fluxgui', in case
            # 'fluxgui' is not installed on a standard system path. We
            # use 'sh' to start the users '/etc/passwd' shell via
            # '$SHELL', so that this will still work if the user
            # changes their shell after the
            # 'autostart/fluxgui.desktop' file is created.
            #
            # See PR #89 for an alternative approach:
            #
            #   https://github.com/xflux-gui/fluxgui/pull/89
            #
            # The escaping of the 'Exec' field is described in
            #
            #   https://developer.gnome.org/desktop-entry-spec/#exec-variables.
            starter_item.set('Exec', r'sh -c "\\"\\$SHELL\\" -c fluxgui"')
            starter_item.set('Icon', 'fluxgui')
            starter_item.set('X-GNOME-Autostart-enabled', 'true')
            starter_item.write()
            self.autostart = True

    def _delete_autostarter(self):
        autostart_file = self._get_autostart_file_path()
        if os.path.isfile(autostart_file):
            os.remove(autostart_file)
            self.autostart = False
